// prisma-generate-schema/dist/src/generator/query/manyQueryArgumentsGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("../generator");
var type_1 = require("graphql/type");
var ManyQueryArgumentsGenerator = /** @class */ (function (_super) {
    __extends(ManyQueryArgumentsGenerator, _super);
    function ManyQueryArgumentsGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ManyQueryArgumentsGenerator.prototype.generate = function (model, args) {
        return {
            where: { type: this.generators.modelWhereInput.generate(model, {}) },
            orderBy: { type: this.generators.modelOrderByInput.generate(model, {}) },
            skip: { type: type_1.GraphQLInt },
            after: { type: type_1.GraphQLString },
            before: { type: type_1.GraphQLString },
            first: { type: type_1.GraphQLInt },
            last: { type: type_1.GraphQLInt }
        };
    };
    return ManyQueryArgumentsGenerator;
}(generator_1.ArgumentsGenerator));
exports.default = ManyQueryArgumentsGenerator;
