// prisma-generate-schema/dist/src/generator/query/modelEdgeGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var type_1 = require("graphql/type");
var generator_1 = require("../generator");
var ModelEdgeGenerator = /** @class */ (function (_super) {
    __extends(ModelEdgeGenerator, _super);
    function ModelEdgeGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ModelEdgeGenerator.prototype.getTypeName = function (input, args) {
        return input.name + "Edge";
    };
    ModelEdgeGenerator.prototype.generateInternal = function (input, args) {
        return new type_1.GraphQLObjectType({
            name: this.getTypeName(input, args),
            fields: {
                node: {
                    type: new type_1.GraphQLNonNull(this.generators.model.generate(input, {})),
                },
                cursor: { type: new type_1.GraphQLNonNull(type_1.GraphQLString) },
            },
        });
    };
    return ModelEdgeGenerator;
}(generator_1.ModelObjectTypeGenerator));
exports.default = ModelEdgeGenerator;
