// prisma-generate-schema/dist/src/generator/query/nodeGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("../generator");
var type_1 = require("graphql/type");
var NodeGenerator = /** @class */ (function (_super) {
    __extends(NodeGenerator, _super);
    function NodeGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NodeGenerator.prototype.getTypeName = function (input, args) {
        return 'Node';
    };
    NodeGenerator.prototype.generateInternal = function (input, args) {
        return new type_1.GraphQLInterfaceType({
            name: this.getTypeName(input, args),
            fields: {
                id: { type: new type_1.GraphQLNonNull(type_1.GraphQLID) }
            }
        });
    };
    return NodeGenerator;
}(generator_1.AuxillaryInterfaceGenerator));
exports.default = NodeGenerator;
