// prisma-generate-schema/dist/src/generator/query/pageInfoGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var type_1 = require("graphql/type");
var generator_1 = require("../generator");
var PageInfoGenerator = /** @class */ (function (_super) {
    __extends(PageInfoGenerator, _super);
    function PageInfoGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PageInfoGenerator.prototype.getTypeName = function (input, args) {
        return 'PageInfo';
    };
    PageInfoGenerator.prototype.generateInternal = function (input, args) {
        return new type_1.GraphQLObjectType({
            name: this.getTypeName(input, args),
            fields: {
                hasNextPage: { type: new type_1.GraphQLNonNull(type_1.GraphQLBoolean) },
                hasPreviousPage: { type: new type_1.GraphQLNonNull(type_1.GraphQLBoolean) },
                startCursor: { type: type_1.GraphQLString },
                endCursor: { type: type_1.GraphQLString },
            },
        });
    };
    return PageInfoGenerator;
}(generator_1.AuxillaryObjectTypeGenerator));
exports.default = PageInfoGenerator;
