// prisma-generate-schema/dist/src/generator/mutation/updateMany/batchPayloadGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var type_1 = require("graphql/type");
var generator_1 = require("../../generator");
var scalarTypeGenerator_1 = require("../../scalar/scalarTypeGenerator");
var BatchPayloadGenerator = /** @class */ (function (_super) {
    __extends(BatchPayloadGenerator, _super);
    function BatchPayloadGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BatchPayloadGenerator.prototype.getTypeName = function (input, args) {
        return 'BatchPayload';
    };
    BatchPayloadGenerator.prototype.generateInternal = function (input, args) {
        return new type_1.GraphQLObjectType({
            name: this.getTypeName(input, args),
            fields: {
                // TODO: Replace with custom long type.
                count: {
                    type: new type_1.GraphQLNonNull(this.generators.scalarTypeGenerator.generate(scalarTypeGenerator_1.TypeIdentifiers.long, {})),
                },
            },
        });
    };
    return BatchPayloadGenerator;
}(generator_1.AuxillaryObjectTypeGenerator));
exports.default = BatchPayloadGenerator;
