// prisma-generate-schema/dist/src/generator/mutation/create/modelCreateManyWithoutRelatedInputGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("../../generator");
var util_1 = require("../../../util/util");
var ModelCreateOneOrManyWithoutRelatedInputGenerator = /** @class */ (function (_super) {
    __extends(ModelCreateOneOrManyWithoutRelatedInputGenerator, _super);
    function ModelCreateOneOrManyWithoutRelatedInputGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ModelCreateOneOrManyWithoutRelatedInputGenerator.prototype.wouldBeEmpty = function (model, args) {
        return !generator_1.TypeFromModelGenerator.hasFieldsExcept.apply(generator_1.TypeFromModelGenerator, [model.fields].concat(generator_1.TypeFromModelGenerator.reservedFields, [args.relatedField.relatedField.name])) &&
            !generator_1.TypeFromModelGenerator.hasUniqueField(model.fields);
    };
    ModelCreateOneOrManyWithoutRelatedInputGenerator.prototype.generateFields = function (model, args) {
        var fields = {};
        if (generator_1.TypeFromModelGenerator.hasFieldsExcept.apply(generator_1.TypeFromModelGenerator, [model.fields].concat(generator_1.TypeFromModelGenerator.reservedFields, [args.relatedField.relatedField.name]))) {
            fields.create = { type: this.maybeWrapList(this.generators.modelCreateWithoutRelatedInput.generate(model, args)) };
        }
        if (generator_1.TypeFromModelGenerator.hasUniqueField(model.fields)) {
            fields.connect = { type: this.maybeWrapList(this.generators.modelWhereUniqueInput.generate(model, args)) };
        }
        return fields;
    };
    return ModelCreateOneOrManyWithoutRelatedInputGenerator;
}(generator_1.RelatedModelInputObjectTypeGenerator));
exports.ModelCreateOneOrManyWithoutRelatedInputGenerator = ModelCreateOneOrManyWithoutRelatedInputGenerator;
// tslint:disable-next-line:max-classes-per-file
var ModelCreateManyWithoutRelatedInputGenerator = /** @class */ (function (_super) {
    __extends(ModelCreateManyWithoutRelatedInputGenerator, _super);
    function ModelCreateManyWithoutRelatedInputGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ModelCreateManyWithoutRelatedInputGenerator.prototype.getTypeName = function (input, args) {
        var field = args.relatedField.relatedField;
        return input.name + "CreateManyWithout" + util_1.capitalize(field.name) + "Input";
    };
    ModelCreateManyWithoutRelatedInputGenerator.prototype.maybeWrapList = function (input) {
        return this.generators.scalarTypeGenerator.wrapList(input);
    };
    return ModelCreateManyWithoutRelatedInputGenerator;
}(ModelCreateOneOrManyWithoutRelatedInputGenerator));
exports.default = ModelCreateManyWithoutRelatedInputGenerator;
