// prisma-generate-schema/dist/src/generator/mutation/create/scalarListCreateInput.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("../../generator");
var type_1 = require("graphql/type");
var ScalarListCreateInput = /** @class */ (function (_super) {
    __extends(ScalarListCreateInput, _super);
    function ScalarListCreateInput() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScalarListCreateInput.prototype.getTypeName = function (input, args) {
        return input.name + "Create" + args.name + "Input";
    };
    ScalarListCreateInput.prototype.generateInternal = function (input, args) {
        return new type_1.GraphQLInputObjectType({
            name: this.getTypeName(input, args),
            fields: {
                set: {
                    type: this.generators.scalarTypeGenerator.wrapList(this.generators.scalarTypeGenerator.generate(args.type, {}))
                }
            }
        });
    };
    return ScalarListCreateInput;
}(generator_1.ModelInputObjectTypeGeneratorBase));
exports.default = ScalarListCreateInput;
