// graphcool-yml/dist/EnvironmentMigrator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Output_1 = require("./Output");
var path = require("path");
var fs = require("fs-extra");
var yaml = require("js-yaml");
var lodash_1 = require("lodash");
var debug = require('debug')('EnvironmentMigrator');
var readMore = ". Read more here: https://goo.gl/3bLCVV";
var EnvironmentMigrator = /** @class */ (function () {
    function EnvironmentMigrator(home, out) {
        if (out === void 0) { out = new Output_1.Output(); }
        this.home = home;
        this.out = out;
        this.graphcoolPath = path.join(this.home, '.graphcool');
        this.graphcoolBackupPath = path.join(this.home, '.graphcool.backup');
        this.rcBackupPath = path.join(this.home, '.graphcoolrc.backup');
        this.rcPath = path.join(this.home, '.graphcoolrc');
    }
    EnvironmentMigrator.prototype.migrate = function () {
        if (fs.pathExistsSync(this.graphcoolPath)) {
            var isFile = fs.lstatSync(this.graphcoolPath).isFile();
            if (isFile) {
                this.migrateGraphcoolFile();
            }
            else {
                this.migrateGraphcoolFolder();
            }
            debug("graphcoolPath " + this.graphcoolPath + " does exists");
        }
        else {
            debug("graphcoolPath " + this.graphcoolPath + " does not exist");
        }
        if (fs.pathExistsSync(this.rcPath)) {
            var isFile = fs.lstatSync(this.rcPath).isFile();
            if (isFile) {
                this.migrateRC();
            }
            else {
                this.migrateRCFolder();
            }
        }
    };
    EnvironmentMigrator.prototype.migrateGraphcoolFile = function () {
        var file = fs.readFileSync(this.graphcoolPath, 'utf-8');
        // only migrate yaml, if it's not json
        try {
            var content = JSON.parse(file);
            this.migrateGraphcoolJsonFile();
        }
        catch (e) {
            try {
                var content = yaml.safeLoad(file);
                this.migrateGraphcoolYamlFile(content);
            }
            catch (e) {
                //
            }
        }
    };
    EnvironmentMigrator.prototype.migrateGraphcoolYamlFile = function (content) {
        this.mergeAndSaveGraphcoolRC({
            'graphcool-1.0': content,
        });
        this.moveGraphcoolFile();
        this.out.log('Its content has been migrated to ~/.graphcoolrc');
    };
    EnvironmentMigrator.prototype.mergeAndSaveGraphcoolRC = function (content) {
        var fileExists = fs.pathExistsSync(this.rcPath);
        var file = fileExists ? fs.readFileSync(this.rcPath, 'utf-8') : null;
        var currentRC = file ? yaml.safeLoad(file) : {};
        currentRC = currentRC.clusters || currentRC.platformToken ? {} : currentRC;
        var newRC = lodash_1.merge(currentRC, content);
        var rcString = yaml.safeDump(JSON.parse(JSON.stringify(newRC)));
        fs.writeFileSync(this.rcPath, rcString);
    };
    EnvironmentMigrator.prototype.migrateGraphcoolJsonFile = function () {
        this.moveGraphcoolFile();
    };
    EnvironmentMigrator.prototype.migrateGraphcoolFolder = function () {
        var configPath = path.join(this.home, '.graphcool/config.yml');
        if (fs.pathExistsSync(configPath)) {
            var file = fs.readFileSync(configPath, 'utf-8');
            try {
                var content = yaml.safeLoad(file);
                this.mergeAndSaveGraphcoolRC({
                    'graphcool-1.0': content,
                });
                this.out.warn("Moved content of ~/.graphcool/config.yml to ~/.graphcoolrc" + readMore);
            }
            catch (e) {
                //
            }
        }
        if (!fs.pathExistsSync(this.graphcoolBackupPath)) {
            fs.moveSync(this.graphcoolPath, this.graphcoolBackupPath);
            this.out.warn("~/.graphcool/ folder found. It has been moved to ~/.graphcool.backup.");
        }
        else {
            this.out
                .warn("~/.graphcool/ folder found. It could not be moved, as ~/.graphcool.backup already exists.\nPlease remove ~/.graphcool by hand to prevent this warning.");
        }
    };
    EnvironmentMigrator.prototype.migrateRC = function () {
        var rc = fs.readFileSync(this.rcPath, 'utf-8');
        var success = false;
        try {
            var content = yaml.safeLoad(rc);
            if (!this.hasCorrectFormat(content)) {
                this.mergeAndSaveGraphcoolRC({
                    'graphcool-framework': content,
                });
                success = true;
            }
            else {
                return;
            }
        }
        catch (e) {
            //
        }
        if (!fs.pathExistsSync(this.rcBackupPath)) {
            fs.writeFileSync(this.rcBackupPath, rc);
            this.out.warn("~/.graphcoolrc file found. It has been moved to ~/.graphcoolrc.backup.");
        }
        else {
            this.out
                .warn("~/.graphcoolrc file found. It could not be moved, as ~/.graphcoolrc.backup already exists.\nPlease remove ~/.graphcoolrc by hand to prevent this warning.");
        }
        if (success) {
            this.out.log("Its content has been migrated to ~/.graphcoolrc");
        }
    };
    EnvironmentMigrator.prototype.migrateRCFolder = function () {
        this.moveRCFolder();
    };
    EnvironmentMigrator.prototype.moveGraphcoolFile = function () {
        if (!fs.pathExistsSync(this.graphcoolBackupPath)) {
            fs.moveSync(this.graphcoolPath, this.graphcoolBackupPath);
            this.out.warn("Old ~/.graphcool file found. It has been moved to ~/.graphcool.backup.");
        }
        else {
            this.out
                .warn("Old ~/.graphcool file found. It could not be moved, as ~/.graphcool.backup already exists.\nPlease remove ~/.graphcool by hand to prevent this warning.");
        }
    };
    EnvironmentMigrator.prototype.moveRCFolder = function () {
        if (!fs.pathExistsSync(this.rcBackupPath)) {
            fs.moveSync(this.rcPath, this.rcBackupPath);
            this.out.warn("~/.graphcoolrc/ folder found. It has been moved to ~/.graphcoolrc.backup/.");
        }
        else {
            this.out
                .warn("~/.graphcoolrc folder found. It could not be moved, as ~/.graphcoolrc.backup already exists.\nPlease remove ~/.graphcoolrc/ by hand to prevent this warning.");
        }
    };
    EnvironmentMigrator.prototype.moveRCFile = function () {
        if (!fs.pathExistsSync(this.rcBackupPath)) {
            fs.moveSync(this.rcPath, this.rcBackupPath);
            this.out.warn("~/.graphcoolrc file found. It has been moved to " + this.rcBackupPath + ".");
        }
        else {
            this.out
                .warn("~/.graphcoolrc file found. It could not be moved, as ~/.graphcoolrc.backup already exists.\nPlease remove ~/.graphcoolrc by hand to prevent this warning.");
        }
    };
    EnvironmentMigrator.prototype.hasCorrectFormat = function (content) {
        if (!content) {
            return false;
        }
        if (Object.keys(content).length > 0) {
            return content['graphcool-1.0'] || content['graphcool-framework'];
        }
    };
    return EnvironmentMigrator;
}());
exports.EnvironmentMigrator = EnvironmentMigrator;
