// prisma-generate-schema/dist/src/generator/schemaGenerator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var generator_1 = require("./generator");
var type_1 = require("graphql/type");
var SchemaGenerator = /** @class */ (function (_super) {
    __extends(SchemaGenerator, _super);
    function SchemaGenerator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SchemaGenerator.prototype.getTypeName = function (input, args) {
        return "__Schema";
    };
    SchemaGenerator.prototype.generate = function (input, args) {
        return new type_1.GraphQLSchema({
            query: this.generators.query.generate(input, {}),
            mutation: this.generators.mutation.generate(input, {}),
            subscription: this.generators.subscription.generate(input, {}),
        });
    };
    return SchemaGenerator;
}(generator_1.SchemaGeneratorBase));
exports.default = SchemaGenerator;
