// prisma-cli-core/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var chalk_1 = require("chalk");
var devPrefix = process.env.ENV === 'DEV' ? 'dev.' : '';
exports.consoleURL = function (token, projectName) {
    return "https://" + devPrefix + "console.graph.cool/token?token=" + token + (projectName ? "&redirect=/" + encodeURIComponent(projectName) : '');
};
// export const playgroundURL = (token: string, projectName: string) =>
function sortByTimestamp(a, b) {
    return a.timestamp < b.timestamp ? -1 : 1;
}
exports.sortByTimestamp = sortByTimestamp;
/**
 * Print a list of [['key', 'value'],...] pairs properly padded
 * @param {string[][]} arr1
 * @param {number} spaceLeft
 * @param {number} spaceBetween
 */
exports.prettyProject = function (p) { return chalk_1.default.bold(p.name) + " (" + p.id + ")"; };
function prettyTime(time) {
    var output = time > 1000 ? (Math.round(time / 100) / 10).toFixed(1) + 's' : time + 'ms';
    return chalk_1.default.cyan(output);
}
exports.prettyTime = prettyTime;
function concatName(name, workspace) {
    var workspaceString = workspace ? workspace + "~" : '';
    return "" + workspaceString + name;
}
exports.concatName = concatName;
