// graphql-schema-linter/lib/rules/defined_types_are_used.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefinedTypesAreUsed = DefinedTypesAreUsed;

var _validation_error = require('../validation_error');

function DefinedTypesAreUsed(context) {
  var ignoredTypes = ['Query', 'Mutatation', 'Subscription'];
  var definedTypes = [];
  var referencedTypes = new Set();

  var recordDefinedType = function recordDefinedType(node) {
    if (ignoredTypes.indexOf(node.name.value) == -1) {
      definedTypes.push(node);
    }
  };

  return {
    ScalarTypeDefinition: recordDefinedType,
    ObjectTypeDefinition: recordDefinedType,
    InterfaceTypeDefinition: recordDefinedType,
    UnionTypeDefinition: recordDefinedType,
    EnumTypeDefinition: recordDefinedType,
    InputObjectTypeDefinition: recordDefinedType,

    NamedType: function NamedType(node) {
      referencedTypes.add(node.name.value);
    },

    Document: {
      leave: function leave(node) {
        definedTypes.forEach(function (node) {
          if (!referencedTypes.has(node.name.value)) {
            context.reportError(new _validation_error.ValidationError('defined-types-are-used', 'The type `' + node.name.value + '` is defined in the ' + 'schema but not used anywhere.', [node]));
          }
        });
      }
    }
  };
}
